src/v/cloud_storage/inventory/ntp_hashes.cc

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cloud_storage/inventory/ntp_hashes.h"
+
+#include "cloud_storage/logger.h"
+#include "container/fragmented_vector.h"
+#include "hashing/xx.h"
+
+#include <seastar/core/fstream.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/util/file.hh>
+
+namespace cloud_storage::inventory {
+ntp_path_hashes::ntp_path_hashes(
+  model::ntp ntp, std::filesystem::path hash_store_path)
+  : _ntp{std::move(ntp)}
+  , _hash_store_path{std::move(hash_store_path)}
+  , _ntp_hashes_path{_hash_store_path / std::string_view{_ntp.path()}}
+  , _rtc{_as}
+  , _ctxlog{cst_log, _rtc, _ntp.path()} {}
+
+ss::future<bool> ntp_path_hashes::load_hashes() {
+    if (co_await ss::file_exists(_ntp_hashes_path.string())) {
+        co_await ss::util::with_file_input_stream(
+          _ntp_hashes_path,
+          [this](auto& stream) { return load_hashes(stream); });
+        _loaded = true;
+        vlog(
+          _ctxlog.trace,
+          "loaded {} hashes from disk, possible collisions {}",
+          _path_hashes.size(),
+          _possible_collisions.size());
+        co_return true;
+    }
+
+    vlog(_ctxlog.warn, "hash file {} not found", _ntp_hashes_path.string());
+    co_return false;
+}
+
+ss::future<> ntp_path_hashes::load_hashes(ss::input_stream<char>& stream) {
+    while (!stream.eof()) {
+        auto chunk_size_buffer = co_await stream.read_exactly(sizeof(size_t));
+        if (chunk_size_buffer.empty()) {
+            break;
+        }
+
+        iobuf c;
+        c.append(std::move(chunk_size_buffer));
+        auto size_parser = iobuf_parser(std::move(c));
+
+        auto chunk_size = serde::read<size_t>(size_parser);
+        iobuf chunk;
+        chunk.append(co_await stream.read_exactly(chunk_size));
+
+        auto chunk_parser = iobuf_parser{std::move(chunk)};
+        auto hashes = serde::read<fragmented_vector<uint64_t>>(chunk_parser);
+        vlog(_ctxlog.trace, "read {} path hashes from disk", hashes.size());
+
+        for (auto hash : hashes) {
+            if (unlikely(_path_hashes.contains(hash))) {
+                _possible_collisions.insert(hash);
+            } else {
+                _path_hashes.insert(hash);
+            }
+        }
+    }
+}
+
+lookup_result ntp_path_hashes::exists(const remote_segment_path& path) const {
+    vassert(_loaded, "lookup without loading hashes from disk");
+    const auto& s = path().string();
+    const auto hash = xxhash_64(s.data(), s.size());
+    if (unlikely(_possible_collisions.contains(hash))) {
+        return lookup_result::possible_collision;
+    }
```

Comment:
similar to Andrew's question. What are the consequences of collisions? Would a viable strategy to be drop any hashes that collide? I would assume that there isn't much that can be done if a collision is detected other than ignore the situation?

Reply:
If a collision is detected the scrubber will make a HEAD request to the bucket, to check if the path for which the collision occurred. 

So the options for the scrubber will be, given each path in the manifest:

* if path exists - do nothing
* if path missing - make a HEAD request to confirm the object is in bucket- the path may have been uploaded after inv. report was generated
* if collision - make a HEAD  request to confirm if the object is present. 

If the HEAD request returns a 404 we register an anomaly (missing object referenced by manifest). 

I think just dropping hashes that collide will also work, since we are taking the same action if the entry is missing or if there is a collision. There is a rare corner case that might happen though:

* three (or any odd number of) path hashes collide
* we add the first one to set
* on the second one we detect a collision and drop the original path, and do not register the second one. now both are going to be marked as missing
* on seeing the third path we now do not know if it is a collision so we add it to our set.

Now if the third path (or any odd entry) is missing in bucket we will miss this during the scrub. So it might be better to track collisions. If we do track them, then the hash where all paths collide serves as a marker that we should make actual HEAD requests for all entries with that hash.